_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+An interesting thing to note here is that just this algorithm by itself could cause the BDP estimation to increase indefinitely; an increase in window will result in sampling more bytes which in turn will cause the window to be increased further. This phenomenon is called buffer-bloat and was discovered by earlier implementations in gRPC-C core and gRPC-Java. The solution to this is to calculate the bandwidth for every sample and check if it is greater than the maximum bandwidth noted so far. If so, only then increase our window sizes. The bandwidth, as we know, can be calculated by dividing the sample by RTT * 1.5 (remember the sample was for one and a half round trips). If the bandwidth doesn't increase with an increase in sampled bytes that's indicative that this change is because of an increased window size and doesn't really reflect the nature of the network itself. 
+
+While running experiments on VMs in different continents we realized that every once in awhile a rogue unnaturally fast ping-ack at the right time (really the wrong time) would cause our window sizes to go up. This happens because such a ping-ack would cause us to notice a decreased RTT and calculate a high bandwidth value. Now if that sample of bytes was greater than 2/3rd of our window then we would increase the window sizes. However, this ping ack was an aberration and shouldn't have changed our perception of the network RTT al together. Therefore, we keep a running average of the RTTs we note weighted by a constant rather than the total number of samples to heed more to recent RTTs and less to the ones in past. It is important because networks might change over time.
+
+During our implementation we experimented with different parameters like instead of putting our window size to twice the bdp sample we put it to 4/3 times the bdp sample. Note although this is more accurate (since the noted bdp was smaller than or equal to 1.5 the real bandwidth and ideally we want to be twice the real bandwidth so multiplying the sample with 4/3 should put us there), it's not the best option to choose since the growth is slow and it takes some tens of RPCs to ramp up.
+
+Given that we're always bound by the flow control of TCP which for most cases is upper bounded at 4MB, we bound the growth of our window sizes by the same number: 4MB.
```

Comment:
This is the default, what happens when the maximum message size increases over 4mb, does our window size increase as well?